### DISCUSSION TITLE
SimpleNetwork behavior with intermediate routers (assertion failed: m_num_connected_buffers < m_port_buffers.size() )

### DISCUSSION THREAD

**Discussion by cristobal-ortega** [gem5/gem5#1803] [Q&A]
I'm trying to build an interconnect similar to the following diagram:
```
 CPU0   CPU1    L30   L31                L31     L30   CPU1   CPU0   
   |     |       |     |                  |       |     |      |      <-- external links
   R     R       R     R      MEM         R       R     R      R
   |     |       |     |       ^          |       |     |      |
   |     |       |     |       |          |       |     |      |
   v     v       v     v       v          ^       ^     ^      ^
   R---->R------>R---->R------>R--------->R------>R---->R----->R   <--- intermediate routers with no external links
   ^     ^       ^     ^                  v       v     v      v
   |     |       |     |                  |       |     |      |
   R     R       R     R                  R       R     R      R
   |     |       |     |                  |       |     |      |      <---external links
  CPU2  CPU3    L32   L33                L33     L32   CPU3   CPU2
```
Note that there are only 4 CPUs and 4 L3s, they are duplicated just for simplicity.

It's a not-closed unidirectional ring with 2 paths: one to go from CPUs/L3s to memory, and another from memory to L3s/CPUs. Hope it's clear.
I based the code on the ring interconnection presented in https://gem5bootcamp.github.io/2024/#03-Developing-gem5-models/08-ruby-network (which I successfully simulated)

With my modifications, I get to the following error:
```Assertion `m_num_connected_buffers < m_port_buffers.size()' failed.```

After looking into the code, it's not clear to me how should I build this intermediate routers in my "ring":
- In the `setup_buffers` function in SimpleNetwork.py, one buffer is appended to a router for each link with that router as destination
- In the `makeInternalLink` function in SimpleNetwork.cc, the dest switch gets an InPort while the src switch gets an OutPort
- In the `addOutPort` function in Switch.cc, the src router checks if the connected buffers are less than the available buffers and, if possible, adds the connection to an available buffer.

So, the intermediate routers coming from memory (right part of the ASCII diagram) in my ring interconnect cannot be modeled since:
- they are destination of one router (the previous router in the ring),
- each router will have a single buffer (or as many as virtual channels per link, let's assume 1 virt channel for simplicity),
- `addOutPort` will try to use 3 buffers (2 CPUs or 2 L3s per router + other intermediate router).

Is my understanding correct? Why this behavior? How can I modeled this interconnect?

**Comment by powerjg** [comment]
You may need to add a buffer per virtual channel. I would trace back the code and see why `m_num_connected_buffers < m_port_buffers.size()`

**Reply by cristobal-ortega** [reply]
Hey Jason, thanks for the quick answer :)

I traced back the code and my understanding is that a buffer per virtual channel is reserved in the router for every link with this router as destination.
Yet, when making the internal link (`makeInternalLink` function) the src router in the link takes 1 buffer per virtual channel.
This works fine if a router has the same inputs and outputs.

Am I missing something?

**Reply by powerjg** [reply]
Yes and no... Clearly there *something* that's missing ;). However, I don't know what is either :)

I'll give you two more pieces of information that are probably not helpful, but this comprises of everything I know.

1. The network code is brittle. Going outside of the topologies that were originally conceived by the authors (back in the mid 2000s) found in configs/topologies has never really been tested.
2. Even my ring network (which was the first time I've tried a bespoke topology) didn't really work as I expected. When I ran in FS mode, there were bugs. So, I don't even think that example is fully "correct."

Basically, the only source of truth is the code, and that's also the only place the knowledge of how networks work exists. Or, to put it another way, YMMV.
